File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.scala
import scala.collection.mutable._

object Solution {
  def canPartition(nums: Array[Int]): Boolean = {
    val sum = nums.sum
    if (sum % 2 == 1) false
    else {
      val target = sum / 2
      val memo: Array[Array[Int]] = Array.fill(nums.length) { Array.fill(target + 1)(-1) }
      helper(nums, 0, sum / 2, memo)
    }
  }

  def helper(
      nums: Array[Int],
      idx: Int,
      target: Int,
      memo: Array[Array[Int]]
  ): Boolean = {
    if (target == 0) true
    else if (idx >= nums.length) false
    else if (memo(idx)(target) != -1) memo(idx)(target) == 1
    else {
      val n = nums(idx)
      val take = if (n <= target) helper(nums, idx + 1, target - n, memo) else false
      val noTake = helper(nums, idx + 1, target, memo)
      memo(idx)(target) = if (take) 1 else 0
      take || noTake
    }
  }
}

File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int n: nums) {
            sum += n;
        }
        if (sum % 2 != 0) return false;
        int n = nums.length;
        int target = sum / 2;
        Boolean[][] memo = new Boolean[n + 1][target + 1];
        return helper(nums, 0, target, memo);
    }
    
    boolean helper(int[] nums, int n, int target, Boolean[][] memo) {
        if (target == 0) return true;
        if (target < 0 || n >= nums.length) return false;
        
        if (memo[n][target] != null) return memo[n][target];
        
        boolean ret = helper(nums, n + 1, target - nums[n], memo) || helper(nums, n + 1, target, memo);
        memo[n][target] = ret;
        return ret;
    }
}